src/ray/stats/metric_exporter.h

```diff
@@ -0,0 +1,88 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RAY_METRIC_EXPORTER_H
+#define RAY_METRIC_EXPORTER_H
+#include "absl/memory/memory.h"
+#include "opencensus/stats/stats.h"
+#include "opencensus/tags/tag_key.h"
+
+#include "ray/stats/metric.h"
+#include "ray/stats/metric_exporter_client.h"
+#include "ray/util/logging.h"
+#include "ray/util/util.h"
+
+namespace ray {
+namespace stats {
+
+/// Main function of metric exporter is collecting indicator information from
+/// opencensus data view, and finally sends it to the remote ( for example
+/// send metrics to gcs service through rpc. How to use it? Register metrics
+/// exporter after raylet main thread launched. Actually its exporter
+///  has never beed used in raylet, we will enable it later.
+class MetricExporter final : public opencensus::stats::StatsExporter::Handler {
+ public:
+  explicit MetricExporter(std::shared_ptr<MetricExporterClient> metric_exporter_client)
+      : metric_exporter_client_(metric_exporter_client) {}
+  ~MetricExporter() = default;
+  static void Register(std::shared_ptr<MetricExporterClient> metric_exporter_client) {
+    opencensus::stats::StatsExporter::RegisterPushHandler(
+        absl::make_unique<MetricExporter>(metric_exporter_client));
+  }
+
+  void ExportViewData(
+      const std::vector<std::pair<opencensus::stats::ViewDescriptor,
+                                  opencensus::stats::ViewData>> &data) override;
+
+ private:
+  template <class DTYPE>
+  /// Extract raw data from view data, then metric exporter clients can use them
+  /// in points schema.
+  /// \param view_data, raw data in map
+  /// \param metric_name, metric name of view data
+  /// \param keys, metric tags map
+  /// \param points, memory metric vector instance
+  void ExportToPoints(const opencensus::stats::ViewData::DataMap<DTYPE> &view_data,
+                      const std::string &metric_name, std::vector<std::string> &keys,
+                      MetricPoints &points) {
+    for (const auto &row : view_data) {
+      std::unordered_map<std::string, std::string> tags;
+      for (size_t i = 0; i < keys.size(); ++i) {
+        tags[keys[i]] = row.first[i];
+      }
+      // Current timestamp is used for point not view data time.
+      MetricPoint point{.metric_name = metric_name,
+                        .timestamp = current_sys_time_ms(),
+                        .value = static_cast<double>(row.second),
+                        .tags = tags};
+      RAY_LOG(DEBUG) << "Metric name " << metric_name << ", value " << point.value;
+      points.push_back(std::move(point));
+      if (points.size() >= kMaxBatchSize) {
+        RAY_LOG(DEBUG) << "Point size : " << points.size();
+        metric_exporter_client_->ReportMetrics(points);
```

Comment:
is `ReportMetrics` pushing metrics to the data sink? If so, do we have a plan to make it asynchronous?

Reply:
> is `ReportMetrics` pushing metrics to the data sink? If so, do we have a plan to make it asynchronous?


The client can decide whether to support asynchronous sending.
Hot about we assume it's default sync?

Reply:
Is reporting happening in a different thread? If so, I think it is fine to be synchronous.

Reply:
> Is reporting happening in a different thread from a main thread? If so, I think it is fine to make it synchronous by default.

Reporting is in another thread of opencensus. Stats handler will create a new timer thread for every exporter.